File: frontend/src/types/portfolio.ts
export enum Region {
  US = "US",
  Developed = "Developed",
  Emerging = "Emerging",
}

export enum Ticker {
  DFUS = "DFUS",
  DFAI = "DFAI",
  AVEM = "AVEM",
  AVUV = "AVUV",
  AVDV = "AVDV",
  AVES = "AVES",
}

export interface Equity {
  ticker: Ticker;
  fractional: boolean;
  share_price: number;
  market_loading: number;
  size_loading: number;
  value_loading: number;
  profitability_loading: number;
  investment_loading: number;
  region: Region;
}

export interface Position {
  ticker: Ticker;
  equity: Equity;
  value: number;
  shares: number;
  current_proportion: number;
  target_proportion: number;
  drift: number;
}

export interface Portfolio {
  total_value: number;
  active_share: number;
  positions: Position[];
  market_loading: number;
  size_loading: number;
  value_loading: number;
  profitability_loading: number;
  investment_loading: number;
  target_market_loading: number;
  target_size_loading: number;
  target_value_loading: number;
  target_profitability_loading: number;
  target_investment_loading: number;
}

export interface RegionalDistribution {
  region: Region;
  current: number;
  target: number;
}

export interface FactorLoadingRow {
  factor: string;
  loading: number;
  target_loading: number;
  premium: number;
  portfolio_premium: number;
}

export interface ExpectedReturns {
  nominal_arithmetic: number;
  real_arithmetic: number;
  nominal_geometric: number;
  real_geometric: number;
  assumptions: {
    inflation: number;
    vol: number;
    rf: number;
  };
}

export interface FactorAnalysis {
  loadings: FactorLoadingRow[];
  excess_premium: number;
  expected_returns: ExpectedReturns;
}

export interface RebalanceAdjustment {
  ticker: string;
  current_shares: number;
  current_value: number;
  target_value: number;
  adjustment_shares: number;
  adjustment_value: number;
  adjustment_error: number;
  final_shares: number;
  final_value: number;
  price: number;
  fractional: boolean;
}

export interface RebalanceResult {
  adjustments: RebalanceAdjustment[];
  total_infusion: number;
  whole_share_error: number;
}

File: backend/app/schemas/portfolio.py
from typing import List, Dict
from pydantic import BaseModel

from ..core.models import Position, Region, RegionalDistribution


class PortfolioResponse(BaseModel):
    total_value: float
    active_share: float
    positions: List[Position]
    market_loading: float
    size_loading: float
    value_loading: float
    profitability_loading: float
    investment_loading: float
    target_market_loading: float
    target_size_loading: float
    target_value_loading: float
    target_profitability_loading: float
    target_investment_loading: float


class RegionalDistributionResponse(BaseModel):
    distributions: List[RegionalDistribution]


class FactorLoadingRow(BaseModel):
    factor: str
    loading: float
    target_loading: float
    premium: float
    portfolio_premium: float


class ExpectedReturns(BaseModel):
    nominal_arithmetic: float
    real_arithmetic: float
    nominal_geometric: float
    real_geometric: float
    assumptions: Dict[str, float]  # inflation, vol, rf


class FactorAnalysisResponse(BaseModel):
    loadings: List[FactorLoadingRow]
    excess_premium: float
    expected_returns: ExpectedReturns


class RebalanceRequest(BaseModel):
    infusion: float


class RebalanceAdjustment(BaseModel):
    ticker: str
    current_shares: float
    current_value: float
    target_value: float
    adjustment_shares: float
    adjustment_value: float
    adjustment_error: float
    final_shares: float
    final_value: float
    price: float
    fractional: bool


class RebalanceResponse(BaseModel):
    adjustments: List[RebalanceAdjustment]
    total_infusion: float
    whole_share_error: float
